src/hotspot/share/opto/loopnode.hpp

```diff
@@ -258,9 +258,13 @@ class CountedLoopNode : public BaseCountedLoopNode {
   int   stride_con() const;
 
   // Match increment with optional truncation
-  static Node*
-  match_incr_with_optional_truncation(Node* expr, Node** trunc1, Node** trunc2, const TypeInteger** trunc_type,
-                                      BasicType bt);
+  struct TruncatedIncrement {
+    Node* incr = nullptr;
+    Node* trunc1 = nullptr;
+    Node* trunc2 = nullptr;
+    const TypeInteger* trunc_type = nullptr;
+  };
+  static TruncatedIncrement match_incr_with_optional_truncation(Node* expr, BasicType bt);
```

Comment:
Just a drive-by comment: this function and some others that you created are returning a value by copy, for performance reasons it may be better to return a reference or even a pointer as usually is the case in HotSpot.

Reply:
I'm not sure if this is always the case but please correct me if I'm wrong: a cpp compiler already optimizes these kinds of patterns and structs are only allocated on the caller stack frame with a pointer to it passed into callee as an argument (or even leaving fields in registers if the struct is small enough). No copying is actually happening.

I agree this *syntactically looks costly*, please let me know if you find it necessary to switch to passing in a struct reference as an argument.